File: tools/tilemap.py
import struct
from congraph import FauxStream

SYMBOLS         = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-[];'\,./`"
FLIPPED_SYMBOLS = 'abcdefghijklmnopqrstuvwxyz!@#$%^&*()_{}:"|<>?~'

def decode(stream, no_plus, width = 32):
    tilemap = ""
    symbols = {}

    symbol = 0

    tile = "\0\0"
    while True:
        prev = tile

        tile = stream.read(2)
        if not tile: break
        if not no_plus and tile not in symbols and tile == chr(ord(prev[0]) + 2) + prev[1]:
            tilemap += "+"
        else:
            if tile not in symbols:
                if tile[0] == "\0" and " " not in symbols.values():
                    symbols[tile] = " "
                elif FLIPPED_SYMBOLS == "":
                    symbols[tile] = SYMBOLS[symbol]
                    symbol += 1
                else:
                    unflipped_tile = tile[0] + chr(ord(tile[1]) & ~0x40)
                    symbols[unflipped_tile] = SYMBOLS[symbol]
                    flipped_tile = tile[0] + chr(ord(tile[1]) | 0x40)
                    symbols[flipped_tile] = FLIPPED_SYMBOLS[symbol]
                    symbol += 1
            tilemap += symbols[tile]
    if width == 17: # Dirty hack, sorry. :(
        align = tilemap[:6]
        tilemap = tilemap[6:]
    tilemap = "\n".join(["".join(tilemap[i:i+width]) for i in range(0, len(tilemap), width)])
    if width == 17: # Dirty hack, sorry. :(
        tilemap = align + "\n" + tilemap

    return (tilemap, {v: k for k, v in symbols.items()})

def decompress(stream):
    tilemap = ""
    repeat = "\0\0\0\0"

    while True:
        r = stream.read(2)
        if r == "": break
        mask = struct.unpack("<H", r)[0]
        for i in xrange(16):
            mask <<= 1
            if mask & 0x10000:
                repeat = stream.read(4)
            tilemap += repeat

    return FauxStream(tilemap)

def compress(stream, out):
    done = False
    while True:
        mask = 0
        last_dword = "\0\0\0\0"
        current_data = ""
        for i in xrange(16):
            dword = stream.read(4)
            if dword == "":
                done = True
                break
            mask <<= 1
            if dword != last_dword:
                current_data += dword
                last_dword = dword
                mask |= 1
        if done: break
        out.write(struct.pack("<H", mask) + current_data)

def encode(stream, symbols):
    tile = "\0\0"
    data = ""
    while True:
        symbol = stream.read(1)
        if symbol == "\n": continue
        if symbol == "": break
        if symbol == "+":
            tile = chr(ord(tile[0]) + 2) + tile[1]
        else:
            tile = symbols[symbol]
        data += tile
    return data

if __name__ == "__main__":
    import sys

    if "no-flip" in sys.argv:
        SYMBOLS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890-[];'\,./`!@#$%^&*()_{}:\"|<>?~"
        FLIPPED_SYMBOLS = ""
        sys.argv.remove("no-flip")

    width = 32

    if "narrow" in sys.argv:
        width = 17
        sys.argv.remove("narrow")

    no_plus = False
    if "no-plus" in sys.argv:
        sys.argv.remove("no-plus")
        no_plus = True

    stream = open(sys.argv[2])

    if sys.argv[1] == "decompress":
        sys.argv[1] = "decode"
        stream = decompress(stream)
    if sys.argv[1] == "decode":
        tilemap, symbols = decode(stream, no_plus, width)
        print tilemap
        print
        for symbol in ' ' + SYMBOLS:
            if symbol not in symbols or (symbol not in tilemap and FLIPPED_SYMBOLS[SYMBOLS.index(symbol)] not in tilemap): continue
            print "%s=%s" % (symbol, symbols[symbol].encode("hex"))
    elif sys.argv[1] in ("encode", "compress"):
        data, symbols_def = stream.read().split("\n\n", 2)
        symbols = {}
        for line in symbols_def.splitlines():
            if len(line) == 0: continue
            find, replace = line.split("=")
            replace = replace.decode("hex")
            symbols[find] = replace
            if find in SYMBOLS:
                flipped_find = FLIPPED_SYMBOLS[SYMBOLS.index(find)]
                if flipped_find not in symbols:
                    symbols[flipped_find] = replace[0] + chr(ord(replace[1]) | 0x40)
        encoded = encode(FauxStream(data), symbols)
        if sys.argv[1] == "compress":
            compress(FauxStream(encoded), open(sys.argv[3], "w"))
        else:
            open(sys.argv[3], "w").write(encoded)

File: tools/tilemap.c
#include <errno.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <string.h>
#include "fileio.h"
#include "rle.h"

/* Since we're using chars as indexes, we're explicitly specifying unsigned chars in this file
   and ignore pointer sign mismatch so we can use string.h functions */
#pragma GCC diagnostic ignored "-Wpointer-sign"


static const unsigned char SYMBOLS[]         = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-[];'\\,./`";
static const unsigned char FLIPPED_SYMBOLS[] = "abcdefghijklmnopqrstuvwxyz!@#$%^&*()_{}:\"|<>?~";

static const unsigned char flip(unsigned char c)
{
    const unsigned char *p = strchr(SYMBOLS, c);
    if (!p) return 0;
    return FLIPPED_SYMBOLS[p - &SYMBOLS[0]];
}

static int encode(FILE *in, FILE *out)
{
    static uint16_t mapping[256] = {0,};
    memset(mapping, 0xFF, sizeof(mapping));
    
    unsigned data_lines = 0;
    
    while (true) {
        const char *line = file_read_line(in);
        data_lines++;
        if (line[0] == '\n') break; // Reached symbol table
        if (line[0] == '\0') { // Reach EOF, no symbol table
            fprintf(stderr, "No symbol table in input\n");
            return 1;
        }
    }
    
    // Build symbol map
    while (true) {
        unsigned char c = 0;
        unsigned tile = 0;
        if (fscanf(in, "%c = %04x\n", &c, &tile) == 1) {
            fprintf(stderr, "Invalid symbol table in input\n");
            return 1;
        }
        if (!c) break;
        mapping[c] = tile;
        unsigned char flipped = flip(c);
        if (flipped && mapping[flipped] == 0xFFFF) {
            mapping[flipped] = tile | 0x40;
        }
    }
    
    fseek(in, 0, SEEK_SET);
    
    // It's safe to assume we can't fail here unless the user modified the file while we were reading it
    uint16_t tile = 0;
    while (data_lines) {
        unsigned char c = 0;
        file_read(in, &c, sizeof(c));
        if (c == '\n') {
            data_lines--;
            continue;
        }
        if (!c) {
            return 1;
        }
        if (c == '+') {
            tile += 0x200;
        }
        else {
            tile = mapping[c];
            if (tile == 0xFFFF) {
                fprintf(stderr, "Undefined symbol %c\n", c);
                return 1;
            }
        }
        uint8_t data[2] = {
            tile >> 8,
            tile,
        };
        file_write(out, data, sizeof(data));
    }
    return 0;
}

int compress(FILE *in, FILE *out)
{
    FILE *encoded = tmpfile();
    int ret = encode(in, encoded);
    if (ret) {
        fclose(encoded);
        return ret;
    }
    fseek(encoded, 0, SEEK_SET);
    ret = rle_compress(encoded, out);
    fclose(encoded);
    return ret;
}

static unsigned char tile_to_symbol(uint16_t tile, uint16_t *mapping)
{
    const unsigned char *symbols = " ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-[];'\\,./`abcdefghijklmnopqrstuvwxyz!@#$%^&*()_{}:\"|<>?~";

    while (*symbols) {
        if (mapping[*symbols] == tile) return *symbols;
        symbols++;
    }
    return 0;
}

int decode(FILE *in, FILE *out, bool narrow, bool no_flip, bool no_plus)
{
    const unsigned char *effective_symbols = no_flip? (unsigned char *)&"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890-[];'\\,./`!@#$%^&*()_{}:\"|<>?~" : SYMBOLS;
    unsigned last_symbol = 0;
    unsigned width = narrow? 17 : 32;
    
    uint16_t tile = 0;
    uint16_t prev = 0;
    
    static uint16_t mapping[256] = {0,};
    memset(mapping, 0xFF, sizeof(mapping));

    unsigned countdown = narrow? 6 : 32; // 6 rather 17 for alignment;
    
    while (true) {
        prev = tile;
        if (!file_read(in, &tile, sizeof(tile))) break;
        tile = ntohs(tile);
        
        if (!no_plus && tile == prev + 0x200 && !tile_to_symbol(tile, mapping)) {
            unsigned char c = '+';
            file_write(out, &c, sizeof(c));
            goto newline;
        }
        
        unsigned char c = tile_to_symbol(tile, mapping);
        if (c) {
            file_write(out, &c, sizeof(c));
            goto newline;
        }
        
        if ((tile >> 8) == 0 && mapping[' '] == 0xFFFF) {
            c = ' ';
        }
        else {
            c = effective_symbols[last_symbol++];
        }
        if (!c) {
            fprintf(stderr, "Too many different tiles\n");
            return 1;
        }
        if (no_flip || c == ' ') {
            mapping[c] = tile;
        }
        else {
            uint16_t unflipped_tile = tile & ~ 0x40;
            mapping[c] = unflipped_tile;
            mapping[flip(c)] = unflipped_tile | 0x40;
            if (tile != unflipped_tile) {
                c = flip(c);
            }
        }
        file_write(out, &c, sizeof(c));

newline:
        if (--countdown == 0) {
            unsigned char c = '\n';
            file_write(out, &c, sizeof(c));
            countdown = width;
        }
    }
    
    unsigned char c = '\n';
    file_write(out, &c, sizeof(c));
    if (mapping[' '] != 0xFFFF) {
        fprintf(out, " =%04x\n", mapping[' ']);
    }
    
    while (*effective_symbols) {
        if (mapping[*effective_symbols] != 0xFFFF) {
            uint16_t tile = ntohs(mapping[*effective_symbols]);
            fprintf(out, "%c=%02x%02x\n", *effective_symbols, tile & 0xff, tile >> 8);
        }
        effective_symbols++;
    }
    
    return 0;
}

int decompress(FILE *in, FILE *out, bool narrow, bool no_flip, bool no_plus)
{
    FILE *decompressed = tmpfile();
    int ret = rle_decompress(in, decompressed);
    if (ret) {
        fclose(decompressed);
        return ret;
    }
    fseek(decompressed, 0, SEEK_SET);
    ret = decode(decompressed, out, narrow, no_flip, no_plus);
    fclose(decompressed);
    return ret;
}

int main(int argc, char **argv)
{
    bool narrow = false;
    bool no_flip = false;
    bool no_plus = false;
    
    for (unsigned i = 2; i < argc; i++) {
        if (strcmp(argv[i], "--narrow") == 0) {
            narrow = true;
            goto remove;
        }
        if (strcmp(argv[i], "--no-flip") == 0) {
            no_flip = true;
            goto remove;
        }
        if (strcmp(argv[i], "--no-plus") == 0) {
            no_plus = true;
            goto remove;
        }
        continue;
remove:
        for (unsigned j = i; j < argc - 1; j++) {
            argv[j] = argv[j + 1];
        }
        argc--;
        i--;
    }
    
    if (argc != 4) {
usage:
        fprintf(stderr, "Usage: %s encode|compress INPUT OUTPUT\n"
                        "       %s decode|decompress [--narrow] [--no-flip] [--no-plus] INPUT OUTPUT\n", argv[0], argv[0]);
        return 1;
    }
    
    FILE *input = file_open(argv[2], "rb");
    FILE *output = file_open(argv[3], "wb");
    
    int ret = 0;
    
    if (strcmp(argv[1], "encode") == 0) {
        ret = encode(input, output);
    }
    else if (strcmp(argv[1], "compress") == 0) {
        ret = compress(input, output);
    }
    else if (strcmp(argv[1], "decode") == 0) {
        ret = decode(input, output, narrow, no_flip, no_plus);
    }
    else if (strcmp(argv[1], "decompress") == 0) {
        ret = decompress(input, output, narrow, no_flip, no_plus);
    }
    else {
        goto usage;
    }

    fclose(input);
    fclose(output);
    
    if (ret) {
        unlink(argv[3]);
    }
    return ret;
}
